Fixes for Saving throw, Ability, Initiative, Attacks, features and Feats

dmportella opened the pull request:
First time doing a fix like this, hmmm there was not much we could do (that i could see) with the fact that DNDBEYOND team removed the class names from the panes saving throw and ability panes. 

So .... 

To minimise the code changes I added a class name that doesn't exist, I changed the "ct" to "b20" and modified all the code to support that, Including the injection on the document modified event.

The code i pushing works on chrome, I have not tested on other browsers.

If there is any problems or you wish not accept this pull request, it is oke, "I am not married to it".

I would not normally make changes like this, just havent had much practice with this front end stuff is a while.

Some other things can be optimised also.

Fixes; https://github.com/kakaroto/Beyond20/issues/1140 and https://github.com/kakaroto/Beyond20/issues/1145 and https://github.com/kakaroto/Beyond20/issues/1146 and https://github.com/kakaroto/Beyond20/issues/1148 and https://github.com/kakaroto/Beyond20/issues/1149 and on https://github.com/kakaroto/Beyond20/issues/1150 and https://github.com/kakaroto/Beyond20/issues/1152 and https://github.com/kakaroto/Beyond20/issues/1153

dmportella commented:
initiative is also broken looking at that at the moment

dmportella commented:
added fix initiative it works properly in desktop and mobile (kept the same as before) but also fixed for tablet (more or less still some issue with missing modifier on the text box)

dmportella commented:
some of the attacks like unarmed strike is not working, I will check that too and add a fix for it

dmportella commented:
added fix for unarmed strike and vampiric bite also optimised some things ont he code i wrote

dmportella commented:
all the changes here should fix the issues with beyond20 that i could find, if there is more I can try and fix it

mjgiarlo commented:
I tested this branch just now and found that it fixed the problem I was seeing with initiative rolls not being sent over to Roll20. 🍻

ja-ry commented:
+1 build works on Edge when used with Roll20, the following worked with proposed changes:
- Saves
- Initiative
- Unarmed Strikes
- Spell Display in VTT
- Pet Extras

PhilBuescher commented:
Works on chrome. Had to add "arms of" to the const array ACTION for my arms of the astral self to work.

kakaroto commented:
Hey! Just dropping a quick message to say a big big thank you to @dmportella for the PR. I've seen it and I've started a partial review last week and been trying to finish reviewing it every single day since, but things have been extremely chaotic on my side unfortunately. I might not be able to get to it until next week :(
One thing I wanted to say though is that I like your approach. I would have kept `ct-` instead of switching it to `b20-` for the panel names, just so it remains consistent and less code changes needed (in case one gets missed), but that's a minor thing.
The main issue I noticed when reviewing the code though is the hardcoded list of actions, which I don't think will scale well. Instead, I noticed (I don't remember what it was, that was when I reviewed last week and I have about 5 minutes to finish this message and rush out of the house) that the side panel has divs with classes that are indicator of actions, I think it was something like `ddbc-action-name` for example, or something like that, check it out. I think it'd be more stable to look for that class name inside the side panel, and if it's found, then apply the class to the toplevel div to mark it as an action pane, instead of relying on a hardcoded list of actions.. Same technique might work for saving throws/abilities/initiative/etc... (wondering if someone has chrome autotranslate enabled on the site, it might not match the textContent of "saving...").
That's all I can say for now, thanks again and big apologies for my lack of availability, and I promise to check it out, review/merge and make a release as soon as I'm able!
❤️

dmportella commented:
> Hey! Just dropping a quick message to say a big big thank you to @dmportella for the PR. I've seen it and I've started a partial review last week and been trying to finish reviewing it every single day since, but things have been extremely chaotic on my side unfortunately. I might not be able to get to it until next week :( One thing I wanted to say though is that I like your approach. I would have kept `ct-` instead of switching it to `b20-` for the panel names, just so it remains consistent and less code changes needed (in case one gets missed), but that's a minor thing. The main issue I noticed when reviewing the code though is the hardcoded list of actions, which I don't think will scale well. Instead, I noticed (I don't remember what it was, that was when I reviewed last week and I have about 5 minutes to finish this message and rush out of the house) that the side panel has divs with classes that are indicator of actions, I think it was something like `ddbc-action-name` for example, or something like that, check it out. I think it'd be more stable to look for that class name inside the side panel, and if it's found, then apply the class to the toplevel div to mark it as an action pane, instead of relying on a hardcoded list of actions.. Same technique might work for saving throws/abilities/initiative/etc... (wondering if someone has chrome autotranslate enabled on the site, it might not match the textContent of "saving..."). That's all I can say for now, thanks again and big apologies for my lack of availability, and I promise to check it out, review/merge and make a release as soon as I'm able! ❤️

Of course it is no problem, I though it would be the case that you were busy. <3

Yeah the the `ct` change to `b20`, i though it would be better not to use a dndbeyond class name that might exist again, thats why i added b20 instead in case it could break anything if they added them again, I can revert back to ct its no problem.

regarding the action list yeah i was not happy about it, but i dont know the code base well enough so I did that just to get thing working until I had a better way of doing it.

I will check and see, what classes are available and I will make any changes possible.

Hope things quiet down for you soon.

dmportella commented:
@kakaroto I was able to make the changes to everything besides initiative, specially for mobile, there is nothing no class that i can uniquely identify initiative.

But all others are now better and dont rely on any text. I will test a few things and then push the change to the pull request

dmportella commented:
@kakaroto Ok I got all the changes done, found a way to get initiative to work by using a simple regex on "word (-/+number)" and it works pretty well on everything. Everything is using classes now instead of any text or words. I tested Mobile and Desktop

d-wierdsma commented:
First off, thanks so much for putting effort into this @dmportella it's super appreciated!!

One thing that is missing from this PR... When you select a feature from the actions tab in chrome it does show the `Beyond 20` button, however in the Features & Traits tab it does not allow you to post the description into your VTT. 

Added an attached photo as reference
![image](https://github.com/user-attachments/assets/7a84b492-119d-4779-af93-f415ddbbcad1)

dmportella commented:
> First off, thanks so much for putting effort into this @dmportella it's super appreciated!!
> 
> One thing that is missing from this PR... When you select a feature from the actions tab in chrome it does show the `Beyond 20` button, however in the Features & Traits tab it does not allow you to post the description into your VTT.
> 
> Added an attached photo as reference ![image](https://private-user-images.githubusercontent.com/13317139/350793747-7a84b492-119d-4779-af93-f415ddbbcad1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE2NTk5MzAsIm5iZiI6MTcyMTY1OTYzMCwicGF0aCI6Ii8xMzMxNzEzOS8zNTA3OTM3NDctN2E4NGI0OTItMTE5ZC00Nzc5LWFmOTMtZjQxNWRkYmJjYWQxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzIyVDE0NDcxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI5YWRlZTk5ZTE4MmM1MDQ0ODlmZWNmMDY5ZmQ2YzczZGIxZjBlYjAxM2JiYTk1ZjVmODNhZTBjYjFlY2E2NGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.eOVZfo1P1L6dDyr4bVBwnfBvr2p7VN6vg8WwuHrQ7qA)

yeah i didnt start with a list of things that were broken, this is after all the things I experienced broken
I will check see what happened, i wouldnt want to hold this pr because of that, there is many things that might still be broken since dndbeyonds update

dmportella commented:
Ok I have added support back to class features and feats too.
They all work now.

dmportella commented:
@kakaroto I have been checking all tickets raised and testing the functionality that was raised as not working, everything that has been raised is fixed by this pull request. If you need anything else let me know. I will do my best

kakaroto reviewed (approved):
Thank you so so much for the PR and for the patience. I just spent the last few hours reviewing this, testing it and fixing a few things that didn't work. I made comments on the code as I reviewed it, but you don't need to do anything as I've fixed all the issues I found at the same time.
You can check my commits to see what I changed and how (https://github.com/kakaroto/Beyond20/commits/master/)
The one extra thing I did was to add custom pane checks for custom-action and racial-traits, even though that code won't do anything now, but I assume they might remove the pane class from them soon enough, so I preferred to be prepared.
I also found in testing is that the quickroll tooltip was not appearing on mobile format, despite it hooking into the right thing, until I realized it was hidden behind the frame div of the mobile extras info, so I had to bump its z-index from 20 to 10000 before it was visible. 
I think everything works now, but it's past 5AM for me now, so I'll prepare a release tomorrow or during the weekend, depending on availability.
That should give people enough time to stress test this latest version. 
Thanks again!

dmportella commented on src/dndbeyond/content-scripts/character.js:
```diff
@@ -2551,17 +2554,63 @@ function documentModified(mutations, observer) {
+        const initRegex = /\b\w+\s\(([-+]?\d+)\)/g;
+        if (sidebar.length > 0) {
+            const sideBarHeader = sidebar.text().toLowerCase();
+            if (sidebar.closest("svg[class*='ddbc-ability-icon']").length > 0 && sidebar.closest(".ddbc-ability-score-manager").length === 0) {
```
that one is weird it has been working for me but then again i didnt check on other browsers, parent find is way better tho

dmportella commented on src/dndbeyond/content-scripts/character.js:
```diff
@@ -282,18 +281,18 @@ function rollAbilityOrSavingThrow(paneClass, rollType) {
 
 function rollInitiative() {
-    let initiative = $(".ct-initiative-box__value .ddbc-signed-number, .ct-initiative-box__value span[class*='styles_numberDisplay']").text();
+    let initiative = $(".ct-combat__summary-group--initiative .integrated-dice__container, span[class*='styles_value'] span[class*='styles_numberDisplay']").text();
```
not sure here, could have been the initiative was a weird one since all break points things are completely different

dmportella commented:
> Thank you so so much for the PR and for the patience. I just spent the last few hours reviewing this, testing it and fixing a few things that didn't work. I made comments on the code as I reviewed it, but you don't need to do anything as I've fixed all the issues I found at the same time. You can check my commits to see what I changed and how (https://github.com/kakaroto/Beyond20/commits/master/) The one extra thing I did was to add custom pane checks for custom-action and racial-traits, even though that code won't do anything now, but I assume they might remove the pane class from them soon enough, so I preferred to be prepared. I also found in testing is that the quickroll tooltip was not appearing on mobile format, despite it hooking into the right thing, until I realized it was hidden behind the frame div of the mobile extras info, so I had to bump its z-index from 20 to 10000 before it was visible. I think everything works now, but it's past 5AM for me now, so I'll prepare a release tomorrow or during the weekend, depending on availability. That should give people enough time to stress test this latest version. Thanks again!

All is good, I figured there would be a few things to change, I not very familiar with the code base and the features from dndbeyond like digital dice being turned off etc - I would not have thought of that easily. I will run it on my end and if I find anything I will let you know.

Glad to be able to help. Hope all goes well.

dmportella commented:
@kakaroto I got the latest and I am checking on my end the only thing I noticed since the changes made is that not initiative is rolling twice which was not the case before. I will double check again and do a fresh install to see if still happening.
